### PR TITLE
Allow setting the log level via a CLI argument

### PR DESCRIPTION
**Pull request by ricekot**

### CODE REVIEW

**Comment by kingthorin** [comment]
Related to: https://github.com/zaproxy/zaproxy/issues/4513

**Comment by kingthorin** [comment]
I don't see anything wrong with this. But I"m not clear on the implementation, is this using env var values? (ex: Where/how would the user set/override it?).

**Comment by ricekot** [comment]
These can be set by passing flags to the JVM, e.g. 

```
java -Dzap.log.level=debug -jar /zap-2.15.0-SNAPSHOT.jar
```

Maybe we should also update `zap.sh` to accept a flag and append this to the arguments it passes to the JVM.

**Comment by psiinon** [comment]
I think we should have a relatively easy documented way for users to set it, and if that means updating `zap.sh` (and `zap.bat`??) then we should do so.

**Comment by ricekot** [comment]
Updated as discussed in IRC - added a `loglevel` flag that can be used to set the log level.

**Comment by psiinon** [comment]
`-loglevel debug` gives way more than usual, but I guess thats expected?
`-loglevel bad` is ignored - I expected that to fail with some message about an invalid logging level..

Other than that its looking good 😁

**Comment by ricekot** [comment]
> `-loglevel debug` gives way more than usual

We could change it to only set the level for the ZAP classes, and do not change the root log level.

> `-loglevel bad` is ignored - I expected that to fail with some message about an invalid logging level..

That was intentional. I thought a bad log level didn't warrant crashing ZAP, since we have a log level in the log4j2.properties file anyway, but I can tweak this.

**Comment by psiinon** [comment]
+1 for only setting the ZAP logging levels (others may disagree?)
We should at least give a warning/error on a bad level otherwise it might be hard for a user to debug a typo?
The help will say which log levels are supported?

File: zap/src/main/java/org/parosproxy/paros/CommandLine.java [hunk 1]
```diff
@@ -599,6 +609,15 @@ public boolean isNoStdOutLog() {
         return noStdOutLog;
     }
 
+    /**
+     * Returns the specified log level argument if valid, or {@code null}.
```
Comment (thc202) [inline zap/src/main/java/org/parosproxy/paros/CommandLine.java:613]: The "if valid" could be removed now.

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties [hunk 2]
```diff
@@ -1107,7 +1107,7 @@ cfu.warn.nolaunch = The latest ZAP release: {0} has been downloaded to\n{1}\nYou
...
...specified directory instead of the default one\n\t-installdir <dir>        Overrides the code that detects where ZAP has been installed with the specified directory\n\t-h                       Shows all of the command line options available, including those added by add-ons\n\t-help                    The same as -h\n\t-newsession <path>       Creates a new session at the given location\n\t-session <path>          Opens the given session after starting ZAP\n\t-lowmem                  Use the database instead of memory as much as possible - this is still experimental\n\t-experimentaldb          Use the experimental generic database code, which is not surprisingly also still experimental\n\t-nostdout                Disables the default logging through standard output\n\t-loglevel <level>        Sets the log level, overriding the values specified in the log4j2.properties file in the home directory\n\t-sbomzip <path>          Creates a zip file containing all of the available SBOMs\n\t-suppinfo                Reports support info to the command line and exits\n\t-silent                  Ensures ZAP does not make any unsolicited requests, including 'check for updates'\nAdd-on options:\n
```
Comment (thc202) [inline zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties:1110]: Just noticed the duplicated resource name (`cmdline.help =`).

Comment (kingthorin) [reply zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties:1110]: Good catch!

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by kingthorin — APPROVED** [review]

**Pull request merged**